File: gfg/2024_8_august/c++/04_n_meetings_in_one_room.cpp
// TIME COMPEXITY
// O(nlogn)

// SPACE COMPEXITY
// O(n)

// CODE

class Solution
{
    public:
    //Function to find the maximum number of meetings that can
    //be performed in a meeting room.
    static bool cmp(pair<int,int> a, pair<int,int> b){
        return a.second < b.second;
    }
    int maxMeetings(int n, int start[], int end[])
    {
        vector<pair<int,int>> temp;
        for(int i = 0; i < n; i++){
            temp.push_back({start[i], end[i]});
        }
        
        sort(temp.begin(), temp.end(), cmp);
        
        int cnt = 1;
        int ansEnd = temp[0].second;
        
        for(int i = 1; i < n; i++){
            if(temp[i].first > ansEnd){
                cnt++;
                ansEnd = temp[i].second;
            }
        }
        
        return cnt;
    }
};

File: gfg/2024_8_august/python/04_n_meetings_in_one_room.py
# CODE

class Solution:
    def maximumMeetings(self, n, start, end):
        meetings = sorted(zip(start, end), key=lambda x: x[1])
        count = 1
        last_end_time = meetings[0][1]
      
        for i in range(1, n):
            current_start_time, current_end_time = meetings[i]
            if current_start_time > last_end_time:
                count += 1
                last_end_time = current_end_time
       
        return count